server/fishtest/api.py

```diff
@@ -247,62 +241,9 @@ def stats(self):
     def spsa(self):
         return self.request_body.get("spsa", {})
 
-    def get_flag(self):
-        def get_country_code(ip):
-            # https://ipwhois.io/ - free 10k/month requests, response examples:
-            # {'success': True, 'country_code': 'US'}
-            # {'success': False, 'message': 'Invalid IP address'}
-            # {'success': False, 'message': 'Reserved range'}
-
-            code = None  # required by exception
-            req = "https://ipwho.is/" + ip + "?fields=country_code,success,message"
-            try:
-                res = requests.get(req, timeout=1.0)
-                res.raise_for_status()  # also catch return codes >= 400
-            except Exception as e:
-                print("Exception checking GeoIP for {}:\n".format(ip), e, sep="")
-            else:
-                res = res.json()
-                if res.get("success"):
-                    code = res.get("country_code")
-                else:
-                    print(
-                        "Failed GeoIP check for {}: {}".format(ip, res.get("message"))
-                    )
-            return code
-
-        def get_flag_cc(self, ip):
-            db_flag = self.request.userdb.flag_cache.find_one({"ip": ip})
-            if not db_flag:
-                db_flag = {
-                    "ip": ip,
-                    "country_code": get_country_code(ip),
-                    "flag_data_checked_at": datetime.utcnow(),
-                }
-                if db_flag["country_code"] is not None:
-                    self.request.userdb.flag_cache.insert_one(db_flag)
-            return db_flag["country_code"]
-
-        ip = self.request.remote_addr
-        now_dt = datetime.utcnow()
-        clean_flag_data = {"cc": None, "dt": now_dt}
-        # Create a flag_data for a new ip and insert it in flag_cache and db
-        # limit race for workers with the same ip: db and web request can be slow
-        if ip not in flag_cache:
-            flag_cache[ip] = clean_flag_data
-            flag_cache[ip]["cc"] = get_flag_cc(self, ip)
-        # Update the flag_data after a timeout if the ip has not a country code
-        # (eg dev worker with private ip) to preserve the free geoip requests
-        try:
-            ip_cc = flag_cache[ip]["cc"]
-            ip_dt = flag_cache[ip]["dt"]
-        except KeyError:
-            pass
-        else:
-            if ip_cc is None and (now_dt - ip_dt).total_seconds() > 3600 * 4:
-                flag_cache[ip] = clean_flag_data
-                flag_cache[ip]["cc"] = get_flag_cc(self, ip)
-        return flag_cache[ip].get("cc")
+    def get_country_code(self):
+        country_code = self.request.headers.get("X-Country-Code")
+        return "?" if not country_code or country_code == "ZZ" else country_code
```

Comment:
This is working code, so just nitpicking here (aka trolling):
If `get("X-Country-Code")` can be Falsy (`None`, ""):
```python
return "?" if country_code in (None, "", "ZZ") else country_code
```
otherwise:
```python
return "?" if country_code in (None, "ZZ") else country_code
```

Reply:
HTTP headers cannot and will not have an empty value, therefore I'll adopt the second line.